agent/engine/task_manager.go

```diff
@@ -461,59 +474,88 @@ func (mtask *managedTask) progressContainers() {
+		if !transition.actionRequired {
+			mtask.handleContainerChange(dockerContainerChange{cont, DockerContainerChangeEvent{Status: transition.nextState}})
 			continue
 		}
-		transitions[cont.Name] = nextState
-		go transitionFunc(cont, nextState)
+		transitions[cont.Name] = transition.nextState
+		go transitionFunc(cont, transition.nextState)
 	}
 
-	return anyCanTransition, transitions
+	return anyCanTransition, transitions, reasons
 }
 
 type containerTransitionFunc func(container *api.Container, nextStatus api.ContainerStatus)
 
 // containerNextState determines the next state a container should go to.
-// It returns: The state it should transition to, a bool indicating whether any
-// action is required, and a bool indicating whether a known status change is
-// possible.
+// It returns a transition struct including the information:
+// * container state it should transition to,
+// * a bool indicating whether any action is required
+// * an error indicate why this transition can't happend
 //
-// 'Stopped, false, true' -> "You can move it to known stopped, but you don't have to call a transition function"
-// 'Running, true, true' -> "You can move it to running and you need to call the transition function"
-// 'None, false, false' -> "This should not be moved; it has unresolved dependencies or is complete; no knownstatus change"
+// 'Stopped, false, ""' -> "You can move it to known stopped, but you don't have to call a transition function"
+// 'Running, true, ""' -> "You can move it to running and you need to call the transition function"
+// 'None, false, containerDependencyNotResolved' -> "This should not be moved; it has unresolved dependencies;"
 //
 // Next status is determined by whether the known and desired statuses are
 // equal, the next numerically greater (iota-wise) status, and whether
 // dependencies are fully resolved.
-func (mtask *managedTask) containerNextState(container *api.Container) (api.ContainerStatus, bool, bool) {
-	clog := log.New("task", mtask.Task, "container", container)
+func (mtask *managedTask) containerNextState(container *api.Container) *containerTransition {
 	containerKnownStatus := container.GetKnownStatus()
 	containerDesiredStatus := container.GetDesiredStatus()
-	if containerKnownStatus == containerDesiredStatus {
-		clog.Debug("Container at desired status", "desired", containerDesiredStatus)
-		return api.ContainerStatusNone, false, false
-	}
-	if containerKnownStatus > containerDesiredStatus {
-		clog.Debug("Container past desired status")
-		return api.ContainerStatusNone, false, false
+	if containerKnownStatus >= containerDesiredStatus {
+		seelog.Debugf("Container transitioned, desired status: %s, container: %s", containerDesiredStatus, container.String())
+		return &containerTransition{
+			nextState:      api.ContainerStatusNone,
+			actionRequired: false,
+			reason:         dependencygraph.ErrorContainerTransitioned,
```

Comment:
I don't understand why you're putting this here.  It's not an error to be `==` with desired state or to be `>` desired state if that happened naturally (e.g., a container stopped by itself).

Reply:
The name is a little misleading here, we are just distinguishing different reason for container can't be transitioned.